<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
 <head>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <title>例外を作成する</title>
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <link media="all" rel="stylesheet" type="text/css" href="style.css"/>
 </head>
 <body class="docs"><div class="manualnavbar" style="text-align: center;">
 <div class="prev" style="text-align: left; float: left;"><a href="class.exception.html">Exception</a></div>
 <div class="next" style="text-align: right; float: right;"><a href="exception.getmessage.html">Exception::getMessage</a></div>
 <div class="up"><a href="class.exception.html">Exception</a></div>
 <div class="home"><a href="index.html">PHP Manual</a></div>
</div><hr /><div id="exception.construct" class="refentry">
 <div class="refnamediv">
  <h1 class="refname">Exception::__construct</h1>
  <p class="verinfo">(PHP 5 &gt;= 5.1.0)</p><p class="refpurpose"><span class="refname">Exception::__construct</span> &mdash; <span class="dc-title">例外を作成する</span></p>

 </div>
 
 <div class="refsect1 description" id="refsect1-exception.construct-description">
  <h3 class="title">説明</h3>
  <div class="constructorsynopsis dc-description">
   <span class="modifier">public</span>  <span class="methodname"><strong>Exception::__construct()</strong></span>
    ([ <span class="methodparam"><span class="type">string</span> <code class="parameter">$message</code><span class="initializer"> = &quot;&quot;</span></span>
   [, <span class="methodparam"><span class="type">int</span> <code class="parameter">$code</code><span class="initializer"> = 0</span></span>
   [, <span class="methodparam"><span class="type"><a href="class.exception.html" class="type Exception">Exception</a></span> <code class="parameter">$previous</code><span class="initializer"> = <strong><code>NULL</code></strong></span></span>
  ]]] )</div>

  <p class="para rdfs-comment">
   例外を作成します。
  </p>
 </div>

 
 <div class="refsect1 parameters" id="refsect1-exception.construct-parameters">
  <h3 class="title">パラメータ</h3>
  <p class="para">
   <dl>

    <dt>

     <span class="term"><em><code class="parameter">message</code></em></span>
     <dd>

      <p class="para">
       スローする例外メッセージ。
      </p>
     </dd>

    </dt>

    <dt>

     <span class="term"><em><code class="parameter">code</code></em></span>
     <dd>

      <p class="para">
       例外コード。
      </p>
     </dd>

    </dt>

    <dt>

     <span class="term"><em><code class="parameter">previous</code></em></span>
     <dd>

      <p class="para">
       以前に使われた例外。例外の連結に使用します。
      </p>
     </dd>

    </dt>

   </dl>

  </p>
 </div>

 
 <div class="refsect1 changelog" id="refsect1-exception.construct-changelog">
  <h3 class="title">変更履歴</h3>
  <p class="para">
   <table class="doctable informaltable">
    
     <thead>
      <tr>
       <th>バージョン</th>
       <th>説明</th>
      </tr>

     </thead>

     <tbody class="tbody">
      <tr>
       <td>5.3.0</td>
       <td>
        <em><code class="parameter">previous</code></em> パラメータが追加されました。
       </td>
      </tr>

     </tbody>
    
   </table>

  </p>
 </div>

 
 <div class="refsect1 notes" id="refsect1-exception.construct-notes">
  <h3 class="title">注意</h3>
  <blockquote class="note"><p><strong class="note">注意</strong>: 
   <p class="para">
    <em><code class="parameter">message</code></em> は、バイナリセーフ
    <em class="emphasis">ではありません</em>。
   </p>
  </p></blockquote>
 </div>

 
</div><div id="usernotes">
 <div class="head">
  <h3 class="title">User Contributed Notes</h3>
 </div>
<div id="allnotes">  <a name="112607""></a>
  <div class="note">
   <strong class="user">talksonweb at gmail dot com</strong>
   <a href="#112607" class="date">04-Jul-2013 06:43</a>
   <div class="text">
    <div class="phpcode">
<code><span class="html">
For those that haven't done exception chaining. Here's an example. <br />
<br />
This allows you to add the previous exception to the next one and give yourself detailed information in the end as to what happened. This is useful in larger applications.<br />
<br />
<span class="default">&lt;?php<br />
</span><span class="keyword">function </span><span class="default">theDatabaseObj</span><span class="keyword">(){<br />
&nbsp;&nbsp; &nbsp; if( </span><span class="default">database_object </span><span class="keyword">){<br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; return </span><span class="default">database_object</span><span class="keyword">; <br />
&nbsp;&nbsp; &nbsp; }<br />
&nbsp;&nbsp; &nbsp; else{<br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; throw new </span><span class="default">DatabaseException</span><span class="keyword">(</span><span class="string">"Could not connect to the database"</span><span class="keyword">);<br />
&nbsp;&nbsp; &nbsp; }<br />
}<br />
<br />
function </span><span class="default">updateProfile</span><span class="keyword">( </span><span class="default">$userInfo </span><span class="keyword">){<br />
&nbsp;&nbsp; &nbsp; try{<br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; </span><span class="default">$db </span><span class="keyword">= </span><span class="default">theDatabaseObj</span><span class="keyword">();<br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; </span><span class="default">$db</span><span class="keyword">-&gt;</span><span class="default">updateProfile</span><span class="keyword">();<br />
&nbsp;&nbsp; &nbsp; }<br />
&nbsp;&nbsp; &nbsp; catch( </span><span class="default">DatabaseException $e </span><span class="keyword">){<br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; </span><span class="default">$message </span><span class="keyword">= </span><span class="string">"The user :" </span><span class="keyword">. </span><span class="default">$userInfo</span><span class="keyword">-&gt;</span><span class="default">username </span><span class="keyword">. </span><span class="string">" could not update his profile information"</span><span class="keyword">;<br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; </span><span class="comment">/* notice the '$e'. I'm adding the previous exception&nbsp; to this exception. I can later get a detailed view of <br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp;&nbsp; where the problem began. Lastly, the number '12' is&nbsp; an exception code. I can use this for categorizing my <br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; exceptions or don't use it at all. */ <br />
&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; </span><span class="keyword">throw new </span><span class="default">MemberSettingsException</span><span class="keyword">(</span><span class="default">$message</span><span class="keyword">,</span><span class="default">12</span><span class="keyword">,</span><span class="default">$e</span><span class="keyword">);<br />
&nbsp;&nbsp; &nbsp; }<br />
}<br />
<br />
try{<br />
&nbsp;&nbsp; &nbsp; </span><span class="default">updateProfile</span><span class="keyword">( </span><span class="default">$userInfo </span><span class="keyword">);<br />
}<br />
catch( </span><span class="default">MemberSettingsException $e </span><span class="keyword">){<br />
&nbsp;&nbsp; &nbsp; </span><span class="comment">// this will give all information we have collected above. <br />
&nbsp;&nbsp; &nbsp; </span><span class="keyword">echo </span><span class="default">$e</span><span class="keyword">-&gt;</span><span class="default">getTraceAsString</span><span class="keyword">();<br />
}<br />
</span><span class="default">?&gt;</span>
</span>
</code>
    </div>
   </div>
  </div>
 <div class="foot"></div>
  <a name="110968""></a>
  <div class="note">
   <strong class="user">mattsch at gmail dot com</strong>
   <a href="#110968" class="date">01-Jan-2013 12:08</a>
   <div class="text">
    <div class="phpcode">
<code><span class="html">
Be aware that while $previous is quite useful in providing exception chaining and better traceability, none of the internal php exceptions (e.g. PDOException, ReflectionException, etc) are called internally within php with $previous in mind.&nbsp; <br />
<br />
So if your code throws an exception, recovers from it, then catches one of these internal php exceptions, recovers from it and throws another exception, you will not know the first exception that was thrown when calling getPrevious.<br />
<br />
See: https://bugs.php.net/bug.php?id=63873</span>
</code>
    </div>
   </div>
  </div>
 <div class="foot"></div>
  <a name="88585""></a>
  <div class="note">
   <strong class="user">mario dot mueller at twt dot de</strong>
   <a href="#88585" class="date">30-Jan-2009 11:21</a>
   <div class="text">
    <div class="phpcode">
<code><span class="html">
This would be nice for handling ZendFramework Exceptions to your own framwork or CMS Exceptions:<br />
<br />
<span class="default">&lt;?php<br />
</span><span class="keyword">try<br />
{<br />
&nbsp;&nbsp;&nbsp; </span><span class="default">$oZFResponse </span><span class="keyword">= </span><span class="default">$oZFHttpClient</span><span class="keyword">-&gt;</span><span class="default">request</span><span class="keyword">();<br />
}<br />
catch(</span><span class="default">Zend_Http_Exception $exZFHttpEx</span><span class="keyword">)<br />
{<br />
&nbsp;&nbsp;&nbsp; throw new </span><span class="default">Runtime_</span><span class="keyword">{</span><span class="default">$oZFHttpEx</span><span class="keyword">-&gt;</span><span class="default">getCode</span><span class="keyword">()}</span><span class="default">_Exception</span><span class="keyword">(</span><span class="string">'some ErrorMailing Message'</span><span class="keyword">, </span><span class="default">500</span><span class="keyword">, </span><span class="default">$oZFHttpEx</span><span class="keyword">);<br />
}<br />
<br />
</span><span class="default">?&gt;<br />
</span><br />
Regards,<br />
Mario</span>
</code>
    </div>
   </div>
  </div>
 <div class="foot"></div>
</div>
</div>
<hr /><div class="manualnavbar" style="text-align: center;">
 <div class="prev" style="text-align: left; float: left;"><a href="class.exception.html">Exception</a></div>
 <div class="next" style="text-align: right; float: right;"><a href="exception.getmessage.html">Exception::getMessage</a></div>
 <div class="up"><a href="class.exception.html">Exception</a></div>
 <div class="home"><a href="index.html">PHP Manual</a></div>
</div></body></html>
